pkg/virt-operator/deletion/all.go

```diff
@@ -0,0 +1,278 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2018 Red Hat, Inc.
+ *
+ */
+
+package deletion
+
+import (
+	"fmt"
+
+	extclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	apiclient "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/kubecli"
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+func Delete(kv *v1.KubeVirt, clientset kubecli.KubevirtClient) (int, error) {
+
+	objectsDeleted := 0
+
+	gracePeriod := int64(0)
+	deleteOptions := &metav1.DeleteOptions{
+		GracePeriodSeconds: &gracePeriod,
+	}
+
+	// delete vmimigrations, vmirs, vm, vmi
+	vmimList, err := clientset.VirtualMachineInstanceMigration(metav1.NamespaceAll).List(&metav1.ListOptions{})
+	if err != nil {
+		log.Log.Errorf("Failed to delete vmims: %v", err)
+		return objectsDeleted, err
+	}
+	for _, vmim := range vmimList.Items {
+		clientset.VirtualMachineInstanceMigration(vmim.Namespace).Delete(vmim.Namespace, deleteOptions)
+		if err != nil {
+			log.Log.Errorf("Failed to delete vmim %+v: %v", vmim, err)
+			return objectsDeleted, err
+		} else {
+			objectsDeleted++
+		}
+	}
+
+	rslist, err := clientset.ReplicaSet(metav1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		log.Log.Errorf("Failed to delete vmrss: %v", err)
+		return objectsDeleted, err
+	}
+	for _, vmrs := range rslist.Items {
+		clientset.VirtualMachineInstanceMigration(vmrs.Namespace).Delete(vmrs.Namespace, deleteOptions)
+		if err != nil {
+			log.Log.Errorf("Failed to delete vmrs %+v: %v", vmrs, err)
+			return objectsDeleted, err
+		} else {
+			objectsDeleted++
+		}
+	}
+
+	vmlist, err := clientset.VirtualMachine(metav1.NamespaceAll).List(&metav1.ListOptions{})
+	if err != nil {
+		log.Log.Errorf("Failed to delete vm: %v", err)
+		return objectsDeleted, err
+	}
+	for _, vm := range vmlist.Items {
+		clientset.VirtualMachine(vm.Namespace).Delete(vm.Namespace, deleteOptions)
+		if err != nil {
+			log.Log.Errorf("Failed to delete vm %+v: %v", vm, err)
+			return objectsDeleted, err
+		} else {
+			objectsDeleted++
+		}
+	}
+
+	vmilist, err := clientset.VirtualMachineInstance(metav1.NamespaceAll).List(&metav1.ListOptions{})
+	if err != nil {
+		log.Log.Errorf("Failed to delete vmis: %v", err)
+		return objectsDeleted, err
+	}
+	for _, vmi := range vmilist.Items {
+		// remove finalizer first
+		patchStr := fmt.Sprintf(`{"metadata":{"finalizers":"[]"}}`)
```

Comment:
That we remove the finalizer is only something which we do because we want faster turn-around times during development. VMIs should be allowed to shut down normally. We may still want to keep this hack in our clean script.